lib/ssl/derive.c

```diff
@@ -1,716 +0,0 @@
-            pms = NULL;
-            /* now do the server side--check the triple bypass first */
-            rv = PK11_PrivDecryptPKCS1(srvPrivkey, rsaPmsBuf, &outLen,
-                                       sizeof rsaPmsBuf,
-                                       (unsigned char *)enc_pms.data,
-                                       enc_pms.len);
-            /* if decrypt worked we're done with the RSA test */
-            if (rv == SECSuccess) {
-                *pcanbypass = PR_TRUE;
-                break;
-            }
-            /* check for fallback to double bypass */
-            target = isTLS ? CKM_TLS_MASTER_KEY_DERIVE
-                           : CKM_SSL3_MASTER_KEY_DERIVE;
-            pms = PK11_PubUnwrapSymKey(srvPrivkey, &enc_pms,
-                                       target, CKA_DERIVE, 0);
-            rv = ssl_canExtractMS(pms, isTLS, PR_FALSE, pcanbypass);
-            if (rv == SECSuccess && *pcanbypass == PR_FALSE)
-                goto done;
-            break;
-        }
-
-        /* Check for NULL to avoid double free.
-         * SECItem_FreeItem sets data NULL in secitem.c#265
-         */
-        if (enc_pms.data != NULL) {
-            SECITEM_FreeItem(&enc_pms, PR_FALSE);
-        }
-        for (; (privKeytype == ecKey && (testecdh || testecdhe)) ||
-               (privKeytype == rsaKey && testecdhe);) {
-            CK_MECHANISM_TYPE target;
-            SECKEYPublicKey *keapub = NULL;
-            SECKEYPrivateKey *keapriv;
-            SECKEYPublicKey *cpub = NULL; /* client's ephemeral ECDH keys */
-            SECKEYPrivateKey *cpriv = NULL;
-            SECKEYECParams *pecParams = NULL;
-
-            if (privKeytype == ecKey && testecdhe) {
-                /* TLS_ECDHE_ECDSA */
-                pecParams = &srvPubkey->u.ec.DEREncodedParams;
-            } else if (privKeytype == rsaKey && testecdhe) {
-                /* TLS_ECDHE_RSA */
-                const namedGroupDef *ecGroup;
-                int serverKeyStrengthInBits;
-                int signatureKeyStrength;
-                int requiredECCbits;
-
-                /* find a curve of equivalent strength to the RSA key's */
-                requiredECCbits = PK11_GetPrivateModulusLen(srvPrivkey);
-                if (requiredECCbits < 0)
-                    break;
-                requiredECCbits *= BPB;
-                serverKeyStrengthInBits = srvPubkey->u.rsa.modulus.len;
-                if (srvPubkey->u.rsa.modulus.data[0] == 0) {
-                    serverKeyStrengthInBits--;
-                }
-                /* convert to strength in bits */
-                serverKeyStrengthInBits *= BPB;
-
-                signatureKeyStrength =
-                    SSL_RSASTRENGTH_TO_ECSTRENGTH(serverKeyStrengthInBits);
-
-                if (requiredECCbits > signatureKeyStrength)
-                    requiredECCbits = signatureKeyStrength;
-
-                ecGroup = ssl_GetECGroupWithStrength(NULL, requiredECCbits);
```

Comment:
you can update ssl_GetECGroupWithStrength to assert the socket now. This is the only call that doesn't have a socket.